pkg/controller/deployment/deployment_controller.go

```diff
@@ -438,41 +438,150 @@ func (dc *DeploymentController) syncDeployment(key string) error {
 		revision := d.Spec.RollbackTo.Revision
 		if _, err = dc.rollback(d, &revision); err != nil {
 			return err
 		}
 	}
 
+	if dc.isScalingEvent(d) {
+		return dc.sync(d)
+	}
+
 	switch d.Spec.Strategy.Type {
 	case extensions.RecreateDeploymentStrategyType:
-		return dc.syncRecreateDeployment(d)
+		return dc.rolloutRecreate(d)
 	case extensions.RollingUpdateDeploymentStrategyType:
-		return dc.syncRollingUpdateDeployment(d)
+		return dc.rolloutRolling(d)
 	}
 	return fmt.Errorf("unexpected deployment strategy type: %s", d.Spec.Strategy.Type)
 }
 
-// Updates the status of a paused deployment
-func (dc *DeploymentController) syncPausedDeploymentStatus(deployment *extensions.Deployment) error {
+// sync is responsible for reconciling deployments on scaling events or when they
+// are paused.
+func (dc *DeploymentController) sync(deployment *extensions.Deployment) error {
 	newRS, oldRSs, err := dc.getAllReplicaSetsAndSyncRevision(deployment, false)
 	if err != nil {
 		return err
 	}
-	allRSs := append(controller.FilterActiveReplicaSets(oldRSs), newRS)
 
-	// Sync deployment status
+	dc.scale(deployment, newRS, oldRSs)
+	dc.cleanupOldReplicaSets(oldRSs, deployment)
+
+	allRSs := append(oldRSs, newRS)
 	return dc.syncDeploymentStatus(allRSs, newRS, deployment)
 }
 
+// scale scales proportionally in order to mitigate risk. Otherwise, scaling up can increase the size
+// of the new replica set and scaling down can decrease the sizes of the old ones, both of which would
+// have the effect of hastening the rollout progress, which could produce a higher proportion of unavailable
+// replicas in the event of a problem with the rolled out template. Should run only on scaling events or
+// when a deployment is paused and not during the normal rollout process.
+func (dc *DeploymentController) scale(deployment *extensions.Deployment, newRS *extensions.ReplicaSet, oldRSs []*extensions.ReplicaSet) {
+	allRSs := oldRSs
+	if newRS != nil {
+		allRSs = append(allRSs, newRS)
+	}
+
+	// Paused and recreate deployments may not have a new replica set. Additionally,
+	// in case no old replica sets have pods, the controller should scale the new
+	// replica set to the full count.
+	if newRS != nil && len(controller.FilterActiveReplicaSets(oldRSs)) == 0 {
+		if newRS.Spec.Replicas == deployment.Spec.Replicas {
+			return
+		}
+		if newRS.Annotations == nil {
+			newRS.Annotations = make(map[string]string)
+		}
+		newRS.Annotations[deploymentutil.DesiredReplicasAnnotation] = fmt.Sprintf("%d", deployment.Spec.Replicas)
+		newRS.Annotations[deploymentutil.TotalReplicasAnnotation] = fmt.Sprintf("%d", deployment.Spec.Replicas)
+		glog.V(4).Infof("Scaling new replica set %q from %d to %d", newRS.Name, newRS.Spec.Replicas, deployment.Spec.Replicas)
+		if _, _, err := dc.scaleReplicaSetAndRecordEvent(newRS, deployment.Spec.Replicas, deployment); err != nil {
+			glog.Infof("Cannot scale replica set %q to %d replicas: %v", newRS.Name, deployment.Spec.Replicas, err)
```

Comment:
Looks like this re-enqueues the deployment so it will get tried again, but AFAICT does publish an event in case of failure so for someone to look at in case there is an issue.  This is a broader problem outside the scope of this PR, but something we need to think about.  If this is caused by not being able to write to etcd or network connectivity issues, then there is not a whole lot we can do.  For now add a TODO to figure out if this is something the cluster admin should be notified about.